### ISSUE TITLE
NTLM authentication doesn't work out-of-the-box (test_ntlm fails), probably more configuration is needed

### ISSUE THREAD

**Issue by kvv81** [heimdal/heimdal#737]
Hello Heimdal,

I'm a developer of Vastdata company and we would like to use Heimdal library as GSS API for Kerberos + NTLM authentication with our proprietary SMB server.

I have a problem with running NTLM authentication using examples from the library. Could you please advise if I miss some step in configuration? I have a good understanding how Kerberos is done, but I probably miss details on how to add NTLM support over this.

1. Using Linux machine with CentOS 7.4
2. Installed heimdal-workstation and heimdal-libs and samba packages from rpms.
3. Uploaded source code of heimdal-7.7.0 and configured/built it (./configure + make).
4. Made all necessary steps to configure Kerberos with our ActiveDirectory server:
   - added DC IP to /etc/resolv.conf to set DNS properly
   - edited /etc/krb5.conf to properly set all relevant information (default_realm, [realms] and [domain_realm]).
   - Using 'net' utility from samba, joined my Linux machine to ActiveDirectory with 'CNTOS-VOLODYMYR' machine name:

$ sudo net ads join -U <admin> vast-smb-dev.local
Enter <admin>'s password:
Using short domain name -- VAST-SMB-DEV
Joined 'CNTOS-VOLODYMYR' to dns domain 'vast-smb-dev.local'

- run 'kinit' to get TGT from Kerberos for machine name principal (from keytab file):

[centos@cntos-volodymyr heimdal-7.7.0]$ heimdal-kinit -t /etc/krb5.keytab CNTOS-VOLODYMYR$
[centos@cntos-volodymyr heimdal-7.7.0]$ heimdal-klist
Credentials cache: FILE:/tmp/krb5cc_1000
        Principal: CNTOS-VOLODYMYR$@VAST-SMB-DEV.LOCAL

  Issued                Expires               Principal
Jul  1 10:17:25 2020  Jul  1 20:17:25 2020  krbtgt/VAST-SMB-DEV.LOCAL@VAST-SMB-DEV.LOCAL

- run gssapi_server/gssapi_client examples to test Kerberos5:

[centos@cntos-volodymyr heimdal-7.7.0]$ ./appl/test/gssapi_server -p 6789 -s host
Using mech: Kerberos 5
User is `CNTOS-VOLODYMYR$@VAST-SMB-DEV.LOCAL'
Server is `host/cntos-volodymyr.vast-smb-dev.local@VAST-SMB-DEV.LOCAL'
gss_verify_mic: hej
gss_unwrap: hemligt INT
gss_unwrap: hemligt CONF

[centos@cntos-volodymyr heimdal-7.7.0]$ ./appl/test/gssapi_client -p 6789 -s host -m krb5 cntos-volodymyr

[centos@cntos-volodymyr heimdal-7.7.0]$ heimdal-klist
Credentials cache: FILE:/tmp/krb5cc_1000
        Principal: CNTOS-VOLODYMYR$@VAST-SMB-DEV.LOCAL

  Issued                Expires               Principal
Jul  1 10:17:25 2020  Jul  1 20:17:25 2020  krbtgt/VAST-SMB-DEV.LOCAL@VAST-SMB-DEV.LOCAL
Jul  1 10:18:23 2020  Jul  1 20:17:25 2020  host/cntos-volodymyr.vast-smb-dev.local@VAST-SMB-DEV.LOCAL

Until now everything is fine, Kerberos works as expected. However when I try to use NTLM with GSSAPI:

[centos@cntos-volodymyr heimdal-7.7.0]$ ./appl/test/gssapi_client -p 6789 -s host -m ntlm cntos-volodymyr
**lt-gssapi_client: Unknown mechanism 'ntlm' (spnego, krb5, no-oid)**

Also I'm getting weird error while running 'test_ntlm' example:
1. Edited ./lib/gssapi/test_ntlm.c to use proper domain = "VAST-SMB-DEV.LOCAL", re-compiled examples
2. Running example:
[centos@cntos-volodymyr heimdal-7.7.0]$ ./lib/gssapi/test_ntlm
**lt-test_ntlm: accept_sec_context v2 gss-code: 851968  Miscellaneous failure (see text) -- mech-code: 2529638918 unknown mech-code 2529638918 for mech unknown**

The error happens in this flow:
**#0  krb5_get_init_creds_keytab (context=context@entry=0x605270, creds=creds@entry=0x7fffffffdda0, client=<optimized out>, keytab=0x607b10, start_time=start_time@entry=0,
    in_tkt_service=in_tkt_service@entry=0x0, options=<optimized out>) at init_creds_pw.c:2867
#1  0x00007ffff79bbab4 in get_ccache (id=0x604048, destroy=0x604060, context=0x605270) at ntlm/kdc.c:110**
#2  kdc_alloc (minor=0x7fffffffe06c, ctx=0x6049d8) at ntlm/kdc.c:176
#3  0x00007ffff79b8e92 in _gss_ntlm_accept_sec_context (minor_status=0x7fffffffe06c, context_handle=0x6040b8, acceptor_cred_handle=<optimized out>, input_token_buffer=0x7fffffffe080,
    input_chan_bindings=<optimized out>, src_name=0x7fffffffdfe0, mech_type=0x7fffffffdfd8, output_token=0x7fffffffe090, ret_flags=0x7fffffffdfc8, time_rec=0x0,
    delegated_cred_handle=0x7fffffffdfd0) at ntlm/accept_sec_context.c:115
#4  0x00007ffff79afc78 in gss_accept_sec_context (minor_status=minor_status@entry=0x7fffffffe06c, context_handle=context_handle@entry=0x7fffffffe070,
    acceptor_cred_handle=acceptor_cred_handle@entry=0x0, input_token=input_token@entry=0x7fffffffe080, input_chan_bindings=input_chan_bindings@entry=0x0, src_name=src_name@entry=0x0,
    mech_type=mech_type@entry=0x0, output_token=output_token@entry=0x7fffffffe090, ret_flags=ret_flags@entry=0x0, time_rec=time_rec@entry=0x0,
    delegated_cred_handle=delegated_cred_handle@entry=0x0) at mech/gss_accept_sec_context.c:226
#5  0x0000000000401425 in test_libntlm_v1 (flags=flags@entry=0) at test_ntlm.c:82
#6  0x00000000004011a3 in main (argc=<optimized out>, argv=<optimized out>) at test_ntlm.c:330

Bottom line - GSS_S_FAILURE is returned from gss_accept_sec_context.
It looks like something is not configured properly for NTLM. Please advise.
Thanks!

**Comment by lhoward** [comment]
Heimdal only supports NTLM acceptor authentication using its own KDC. (See `_gss_ntlm_allocate_ctx()` in `lib/gssapi/ntlm/accept_sec_context.c`.) So unfortunately this isn't going to work. You'd either need to plumb Samba into the NTLM library by implementing your own set of `ntlm_server_interface` callbacks, or use another NTLM implementation. The latter is probably easier – see https://github.com/simo5/gssntlmssp for example.

**Comment by kvv81** [comment]
Thank  you very much for clarification, for now it is clear why we can't see NETLOGON/DCERPC logic (needed for NTLM pass-through authentication) inside Heimdal GSSAPI code.

**Comment by lhoward** [comment]
Yes, there is part of the NetLogon code in lib/gssapi/netlogon but the actual daemon was never open sourced. But, Samba will have an alternative. Or if you need something BSD licensed, the DCERPC stack from opensource.apple.com is a good starting point.

**Comment by frozencemetery** [comment]
For what it's worth: gss-ntlmssp ([gssapi/gss-ntlmssp](https://github.com/gssapi/gss-ntlmssp)) relicensed recently to ISC (previously LGPLv3), so it should satisfy any BSD needs.

**Comment by lhoward** [comment]
Ah great, thanks @frozencemetery. Perhaps we should remove the NTLM package from Heimdal and test this one works with it.
